turbopack/crates/turbo-tasks/src/scope.rs

```diff
@@ -1,52 +1,230 @@
+    },
+};
 
-/// A wrapper around [`rayon::Scope`] that preserves the [`turbo_tasks_scope`].
-pub struct Scope<'scope, 'a> {
-    scope: &'a rayon::Scope<'scope>,
-    handle: tokio::runtime::Handle,
-    turbo_tasks: Arc<dyn TurboTasksApi>,
-    span: tracing::Span,
+use parking_lot::Mutex;
+use tokio::{
+    runtime::{Builder, Handle},
+    task::{JoinHandle, block_in_place},
+};
+use tracing::{Instrument, Span};
+
+use crate::{
+    TurboTasksApi,
+    manager::{try_turbo_tasks, turbo_tasks_future_scope},
+};
+
+/// Scope to allow spawning tasks with a limited lifetime.
+///
+/// Dropping this Scope will wait for all tasks to complete.
+pub struct Scope<'scope, 'env: 'scope, R: Send + 'env> {
+    results: &'scope [Mutex<Option<R>>],
+    index: AtomicUsize,
+    futures: Mutex<Vec<JoinHandle<()>>>,
+    handle: Handle,
+    turbo_tasks: Option<Arc<dyn TurboTasksApi>>,
+    span: Span,
+    /// Invariance over 'env, to make sure 'env cannot shrink,
+    /// which is necessary for soundness.
+    ///
+    /// see https://doc.rust-lang.org/src/std/thread/scoped.rs.html#12-29
+    env: PhantomData<&'env mut &'env ()>,
 }
 
-impl<'scope> Scope<'scope, '_> {
-    pub fn spawn<Body>(&self, body: Body)
+impl<'scope, 'env: 'scope, R: Send + 'env> Scope<'scope, 'env, R> {
+    fn new(results: &'scope [Mutex<Option<R>>]) -> Self {
+        Self {
+            results,
+            index: AtomicUsize::new(0),
+            futures: Mutex::new(Vec::with_capacity(results.len())),
+            handle: Handle::current(),
+            turbo_tasks: try_turbo_tasks(),
+            span: Span::current(),
+            env: PhantomData,
+        }
+    }
+
+    /// Spawns a new task in the scope.
+    pub fn spawn<F>(&self, f: F)
     where
-        Body: FnOnce(&Scope<'scope, '_>) + Send + 'scope,
+        F: Future<Output = R> + Send + 'env,
     {
-        let span = self.span.clone();
-        let handle = self.handle.clone();
-        let turbo_tasks = self.turbo_tasks.clone();
-        self.scope.spawn(|scope| {
-            let _span = span.clone().entered();
-            let _guard = handle.enter();
-            turbo_tasks_scope(turbo_tasks.clone(), || {
-                body(&Scope {
-                    scope,
-                    span,
-                    handle,
-                    turbo_tasks,
-                })
-            })
+        let index = self.index.fetch_add(1, Ordering::Relaxed);
+        assert!(index < self.results.len(), "Too many tasks spawned");
+        let result_cell: &Mutex<Option<R>> = &self.results[index];
+
+        let f: Box<dyn Future<Output = ()> + Send + 'scope> = Box::new(async move {
+            let result = f.await;
+            *result_cell.lock() = Some(result);
         });
+        let f: *mut (dyn Future<Output = ()> + Send + 'scope) = Box::into_raw(f);
+        // SAFETY: Scope ensures (e. g. in Drop) that spawned tasks is awaited before the
+        // lifetime `'env` ends.
```

Comment:
Nit: Really, this guarantee is being provided by `scope_and_block`. `Drop` normally doesn't provide any guarantees that it is called.


```suggestion
        // SAFETY: `scope_and_block` ensures that `Scope` is dropped and that spawned tasks are
        // awaited before the lifetime `'env` ends.
```